File: src/main/java/com/shopjoy/service/InventoryService.java
package com.shopjoy.service;

import com.shopjoy.dto.response.InventoryResponse;
import com.shopjoy.exception.DuplicateResourceException;
import com.shopjoy.exception.ResourceNotFoundException;
import com.shopjoy.exception.ValidationException;

import java.util.List;

/**
 * Service interface for Inventory management operations.
 * Handles stock tracking, updates, and low-stock alerts.
 */
public interface InventoryService {
    
    /**
     * Creates inventory record for a new product.
     * 
     * @param productId the product ID
     * @param initialStock the initial stock quantity
     * @param reorderLevel the reorder level
     * @return the created inventory response
     * @throws ValidationException if inventory data is invalid
     * @throws DuplicateResourceException if inventory already exists for product
     */
    InventoryResponse createInventory(Integer productId, int initialStock, int reorderLevel);
    
    /**
     * Retrieves inventory for a specific product.
     * 
     * @param productId the product ID
     * @return the inventory response
     * @throws ResourceNotFoundException if inventory not found
     */
    InventoryResponse getInventoryByProduct(Integer productId);
    
    /**
     * Checks if a product is in stock.
     * 
     * @param productId the product ID
     * @return true if product has stock, false otherwise
     */
    boolean isProductInStock(Integer productId);
    
    /**
     * Checks if a product has sufficient quantity available.
     * 
     * @param productId the product ID
     * @param quantity the required quantity
     * @return true if sufficient stock available, false otherwise
     */
    boolean hasAvailableStock(Integer productId, int quantity);
    
    /**
     * Updates the stock quantity for a product.
     * 
     * @param productId the product ID
     * @param newQuantity the new stock quantity
     * @return the updated inventory response
     * @throws ResourceNotFoundException if inventory not found
     * @throws ValidationException if quantity is negative
     */
    InventoryResponse updateStock(Integer productId, int newQuantity);
    
    /**
     * Adds stock to a product (restock operation).
     * 
     * @param productId the product ID
     * @param quantity the quantity to add
     * @return the updated inventory response
     * @throws ResourceNotFoundException if inventory not found
     * @throws ValidationException if quantity is invalid
     */
    InventoryResponse addStock(Integer productId, int quantity);
    
    /**
     * Removes stock from a product (sale or damage).
     * Validates that sufficient stock is available.
     * 
     * @param productId the product ID
     * @param quantity the quantity to remove
     * @return the updated inventory response
     * @throws ResourceNotFoundException if inventory not found
     * @throws InsufficientStockException if not enough stock available
     * @throws ValidationException if quantity is invalid
     */
    InventoryResponse removeStock(Integer productId, int quantity);
    
    /**
     * Reserves stock for an order (decreases available quantity).
     * Used during order creation to ensure atomicity.
     * 
     * @param productId the product ID
     * @param quantity the quantity to reserve
     * @throws ResourceNotFoundException if inventory not found
     * @throws InsufficientStockException if not enough stock available
     */
    void reserveStock(Integer productId, int quantity);
    
    /**
     * Releases reserved stock (increases available quantity).
     * Used when order is cancelled.
     * 
     * @param productId the product ID
     * @param quantity the quantity to release
     * @throws ResourceNotFoundException if inventory not found
     */
    void releaseStock(Integer productId, int quantity);
    
    /**
     * Retrieves all products with low stock (at or below reorder level).
     * 
     * @return list of low-stock inventory responses
     */
    List<InventoryResponse> getLowStockProducts();
    
    /**
     * Retrieves all out-of-stock products.
     * 
     * @return list of out-of-stock inventory responses
     */
    List<InventoryResponse> getOutOfStockProducts();
    
    /**
     * Updates the reorder level for a product.
     * 
     * @param productId the product ID
     * @param reorderLevel the new reorder level
     * @return the updated inventory response
     * @throws ResourceNotFoundException if inventory not found
     * @throws ValidationException if reorder level is invalid
     */
    InventoryResponse updateReorderLevel(Integer productId, int reorderLevel);

    /**
     * Retrieves inventory for multiple products in a single batch.
     * 
     * @param productIds the list of product IDs
     * @return list of inventory responses
     */
    List<InventoryResponse> getInventoryByProducts(List<Integer> productIds);
}

File: frontend/src/services/inventoryService.js
import api from './api';
import { getAllProducts } from './productService';

// GET /api/v1/inventory/product/{productId} - Get product inventory
export const getInventory = (productId) => api.get(`/inventory/product/${productId}`);

// GET /api/v1/inventory/product/{productId}/in-stock - Check if product is in stock
export const isProductInStock = (productId) => api.get(`/inventory/product/${productId}/in-stock`);

// GET /api/v1/inventory/product/{productId}/available-stock?quantity={quantity} - Check stock availability
export const hasAvailableStock = (productId, quantity) =>
    api.get(`/inventory/product/${productId}/available-stock`, { params: { quantity } });

// PUT /api/v1/inventory/product/{productId}?newQuantity={quantity} - Update stock quantity
export const updateStock = (productId, newQuantity) =>
    api.put(`/inventory/product/${productId}`, null, { params: { newQuantity } });

// PATCH /api/v1/inventory/product/{productId}/add?quantity={quantity} - Add stock
export const addStock = (productId, quantity) =>
    api.patch(`/inventory/product/${productId}/add`, null, { params: { quantity } });

// PATCH /api/v1/inventory/product/{productId}/remove?quantity={quantity} - Remove stock
export const removeStock = (productId, quantity) =>
    api.patch(`/inventory/product/${productId}/remove`, null, { params: { quantity } });

// PATCH /api/v1/inventory/product/{productId}/reserve?quantity={quantity} - Reserve stock
export const reserveStock = (productId, quantity) =>
    api.patch(`/inventory/product/${productId}/reserve`, null, { params: { quantity } });

// PATCH /api/v1/inventory/product/{productId}/release?quantity={quantity} - Release reserved stock
export const releaseStock = (productId, quantity) =>
    api.patch(`/inventory/product/${productId}/release`, null, { params: { quantity } });

// GET /api/v1/inventory/low-stock - Get low stock products
export const getLowStockProducts = () => api.get('/inventory/low-stock');

// GET /api/v1/inventory/out-of-stock - Get out of stock products
export const getOutOfStockProducts = () => api.get('/inventory/out-of-stock');

// PATCH /api/v1/inventory/product/{productId}/reorder-level?reorderLevel={level} - Update reorder level
export const updateReorderLevel = (productId, reorderLevel) =>
    api.patch(`/inventory/product/${productId}/reorder-level`, null, { params: { reorderLevel } });

// GET /api/v1/inventory/products/batch?productIds={ids} - Get inventory for multiple products
export const getInventoryBatch = (productIds) => 
    api.get('/inventory/products/batch', { params: { productIds: productIds.join(',') } });

// GET all inventory items - now uses EntityGraph to fetch products with inventory in single query
export const getAllInventoryItems = async () => {
    const products = await getAllProducts();
    // ProductResponse now includes stockQuantity and reorderLevel via EntityGraph
    const inventoryItems = products.data.map(product => ({
        ...product,
        productId: product.id,
        productName: product.productName,
        stockQuantity: product.stockQuantity || 0,
        reorderLevel: product.reorderLevel || 10,
        reservedQuantity: 0
    }));
    return { data: inventoryItems };
};
